custom/icds_reports/data_pull/data_pulls.py

```diff
@@ -193,3 +201,77 @@ def _dump_consolidated_data(result):
+
+
+class VHSNDMonthlyReport(MonthBasedDataPull):
+    slug = "vhsnd_monthly_report"
+    name = "VHSND monthly report"
+    queries = [
+        VHSNDMonthlyCount
+    ]
+    file_format_required = 'xlsx'
+
+    def post_run(self, data_files):
+        state_results = self._consolidate_data(data_files)
+        return self._dump_consolidated_data(state_results)
+
+    def _consolidate_data(self, data_files):
+        result = defaultdict(dict)
+        state_name_column = 'state_name'
+        vhnsnd_date_column = 'vhsnd_date_past_month'
+        for filename, filestream in data_files.items():
+            filestream.seek(0)
+            reader = csv.DictReader(filestream)
+            for row in reader:
+                state_name = row[state_name_column]
+                vhsnd_date = parse_date(row[vhnsnd_date_column])
+                vhsnd_date = vhsnd_date.strftime('%d/%m/%Y')
+                data_key = (row['state_name'], row['district_name'], row['block_name'],
+                            row['supervisor_name'], row['awc_name'])
+                if data_key in result[state_name]:
+                    result[state_name][data_key].append(vhsnd_date)
+                else:
+                    result[state_name][data_key] = [vhsnd_date]
+
+        return self._format_consolidated_data(result)
+
+    def _format_consolidated_data(self, result):
+        state_results = defaultdict(dict)
+        output_files = defaultdict()
+        self.month_date = datetime.datetime.strptime(self.month, '%Y-%m-%d')
+        num_days = calendar.monthrange(self.month_date.year, self.month_date.month)[1]
+        days = [datetime.date(self.month_date.year, self.month_date.month, day) for day in range(1, num_days + 1)]
+
+        dates = [day.strftime('%d/%m/%Y') for day in days]
+
+        headers = ['State', 'District', 'Block', 'Sector', 'AWC']
+        headers.extend(dates)
+        headers.append('Grand Total')
+        for state_name in result.keys():
+            # constructing and mapping writers to state names
+            wb = openpyxl.Workbook()
+            ws = wb.create_sheet(title=state_name, index=0)
+            ws.append(headers)
+            state_results[state_name] = ws
+            output_files[state_name] = wb
+        for state_name, all_details in result.items():
+            for details, vhsnd_dates in all_details.items():
+                awc_row = list(copy(details))
+                total_count = 0
+                for a_date in dates:
+                    if a_date in vhsnd_dates:
+                        awc_row.append(1)
+                        total_count += 1
+                    else:
+                        awc_row.append('')
+                awc_row.append(total_count)
+                state_results[state_name].append(awc_row)
+        return output_files
+
+    def _dump_consolidated_data(self, result):
+        output_files = defaultdict()
+        for state_name, output_file in result.items():
```

Comment:
```
for state_name, output_file in result.items():
     filename = "vhsnd_monthly_report_{}_{}_{}.xlsx".format(state_name, self.month_date.strftime('%b'), self.month_date.year)
     filestream = io.BytesIO()
     output_file.save(filestream)
     output_files[filename] = filestream
```
This will help you get rid of csv/xlsx logic.

Reply:
thanks for the info. changed it